Pre go-live cleanup of ff docs (#215)
harness/developer-hub: docs/feature-flags/4-ff-sdks/2-client-sdks/7-react-client.md

---
title: React Client SDK Reference
description: This topic explains how to use the Harness Feature Flags SDK in your React Client application.
tags: 
   - feature flag
   - React Client
   - SDK
# sidebar_position: 7
helpdocs_category_id: y1oewjcb0q
helpdocs_is_private: false
helpdocs_is_published: true
---

This topic describes how to use the Harness Feature Flags SDK for your React Client application. 

For getting started quickly, you can use our [sample code from the SDK README](https://github.com/harness/ff-react-client-sdk/blob/main/README.md). You can also [clone](https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository) and run a sample application from the [React Client SDK GitHub Repository](https://github.com/harness/ff-react-client-sdk).

> ## Before You Begin

Reviewer: ```suggestion
## Before you begin
```

* [Getting Started with Feature Flags](../../1-ff-onboarding/2-ff-getting-started/2-getting-started-with-feature-flags.md)
* [Feature Flags Overview](../../1-ff-onboarding/1-cf-feature-flag-overview.md)
* [Client-Side and Server-Side SDKs](../1-sdk-overview/1-client-side-and-server-side-sdks.md)

> * [Communication Strategy Between SDKs and Harness Feature Flags](../1-sdk-overview/2-communication-sdks-harness-feature-flags.md)

Reviewer: Include a "Before you begin" section only if necessary.

## Version

The current version of this SDK is **1.0.0.**

## Prerequisites

To use this SDK, make sure you've: 

* Installed Node.js v12 or a newer version.
* Installed React.js v16.7 or a newer version.

> * [Create a Feature Flag on the Harness Platform](../../2-ff-using-flags/1-ff-creating-flag/4-create-a-feature-flag.md). If you are following along with the SDK README sample code, make sure your flag is called `harnessappdemodarkmode`.

Reviewer: ```suggestion
* [Created a Feature Flag on the Harness Platform](../../2-ff-using-flags/1-ff-creating-flag/4-create-a-feature-flag.md). If you are following along with the SDK README sample code, make sure your flag is called `harnessappdemodarkmode`.
```

> * [Create an SDK key and make a copy of it](../../2-ff-using-flags/1-ff-creating-flag/4-create-a-feature-flag.md#step-3-create-an-sdk-key)

Reviewer: ```suggestion
* [Created an SDK key and made a copy of it](../../2-ff-using-flags/1-ff-creating-flag/4-create-a-feature-flag.md#step-3-create-an-sdk-key)
```

## Install the SDK

First, install the SDK as a dependency in your application. You can install using npm or Yarn: 

### Install using npm

```
npm install @harnessio/ff-react-client-sdk
```
### Install using Yarn

```
yarn add @harnessio/ff-react-client-sdk

> ```

Reviewer: This is a good place to use tabs to present information for different platforms.

Author: Yes I think so too.

## Initialize the SDK

To initialize the React Client SDK, you need to:

1. Import the relevant components.
2. Add your Client SDK key to connect to your Harness Environment.
3. Add a Target that you want to Evaluate against a Feature Flag.
4. (Optional) Configure the SDK options.

> 5. Complete the initialization.

Reviewer: Remove unnecessary capitalization in terms like environment and evaluate.

### Import the components

> Import `FFContextProvider`  from `@harnessio/ff-react-client-sdk` to wrap your application. This allows you to access the Feature Flags service. For example: 

Reviewer: ```suggestion
Import `FFContextProvider` from `@harnessio/ff-react-client-sdk` to wrap your application. This allows you to access the Feature Flags service. For example: 
```

```
import { FFContextProvider } from '@harnessio/ff-react-client-sdk'
```

> ### Add your Client SDK Key

Reviewer: ```suggestion
### Add your Client SDK key
```

> To connect to the correct Environment that you set up on the Harness Platform, you need to add the Client SDK Key from that Environment. Input the Client SDK Key into the `apiKey` parameter, for example:

Reviewer: ```suggestion
To connect to the correct environment that you set up on the Harness Platform, you need to add the Client SDK key from that environment. Input the Client SDK key into the `apiKey` parameter, for example:
```

```
apiKey="YOUR_API_KEY"
```

> ### Add a Target

Reviewer: ```suggestion
### Add a target
```

<details>

> <summary>What is a Target?</summary> 

Reviewer: ```suggestion
<summary>What is a target?</summary> 
```

> Targets are used to control which users see which Variation of a Feature Flag, for example, if you want to do internal testing, you can enable the Flag for some users and not others. When creating a Target, you give it a name and a unique identifier. Often Targets are users but you can create a Target from anything that can be uniquely identified, such as an app or a machine.  

Reviewer: ```suggestion
Targets are used to control which users see which variation of a feature flag, for example, if you want to do internal testing, you can enable the flag for some users and not others. When creating a target, you give it a name and a unique identifier. Often, targets are users but you can create a target from anything that can be uniquely identified, such as an app or a machine.  
```

  </details>

> For more information about Targets, go to [Targeting Users With Flags](../../2-ff-using-flags/4-ff-target-management/3-targeting-users-with-flags.md).

Reviewer: ```suggestion
For more information about targets, go to [Targeting Users With Flags](../../2-ff-using-flags/4-ff-target-management/3-targeting-users-with-flags.md).
```

> To add a Target that you want to Evaluate, build it using `cfTarget` and pass in arguments for the following:

Reviewer: ```suggestion
To add a target that you want to evaluate, build it using `cfTarget` and pass in arguments for the following:
```

|  |  |  |  |
| --- | --- | --- | --- |
| **Parameter** | **Description** | **Required?** | **Example** |

> | identifier | Unique ID for the Target. | Required | `identifier: 'HT_1',` |

Reviewer: ```suggestion
| identifier | Unique ID for the target. | Required | `identifier: 'HT_1',` |
```

> | name | Name for this Target. This does not have to be unique. **Note**: If you don’t provide a value, the name will be the same as the identifier. Read Regex requirements for Target names and identifiers below for accepted characters. | Required | `name: 'Harness_Target_1'` |

Reviewer: ```suggestion
| name | Name for this target. This does not have to be unique. **Note**: If you don’t provide a value, the name will be the same as the identifier. Read Regex requirements for target names and identifiers below for accepted characters. | Required | `name: 'Harness_Target_1'` |
```

> | attributes | Additional data you can store for a Target, such as email addresses or location. These are key/pair values. | Optional | `email: 'demo@harness.io’ location: 'EMEA'` |

Reviewer: ```suggestion
| attributes | Additional data you can store for a target, such as email addresses or location. These are key/pair values. | Optional | `email: 'demo@harness.io’ location: 'EMEA'` |
```

<details>

> <summary> Regex requirements for Target names and identifiers </summary>

Reviewer: ```suggestion
<summary> Regex requirements for target names and identifiers. </summary>
```

**Identifier** 

Regex: `^[A-Za-z0-9.@_-]*$`  
Must consist of only alphabetical characters, numbers, and the following symbols:  
. (period)  
@ (at sign)  
-(dash)  
\_ (underscore)  
  

> The characters can be lowercase or uppercase but cannot include accented letters, for example `Cafe_789`.  

Reviewer: I don't see an accent on the example characters.

  
**Name**
Regex: `^[\\p{L}\\d .@_-]*$`  
  
Must consist of only alphabetical characters, numbers, and the following symbols:  
. (period)  
@ (at sign)  
-(dash)  
\_ (underscore)  
 (space)  
  
The characters can be lowercase or uppercase and can include accented letters, for example `Café_123`.

</details>
 

For example:

```
<FFContextProvider
      apiKey="YOUR_API_KEY"
      target={{
    identifier: 'HT_1',
    name: 'Harness_Target_1'   attributes: { // OPTIONAL: key/value pairs of attributes
           email: 'demo@harness.io',
           location: 'EMEA'
        }
      }}
    >
```
### Configure the SDK

You can configure the following features of the SDK:

|  |  |  | 
| --- | --- | --- | 
| **Name** | **Description** | **Default Value** |

> | baseUrl | The URL used to fetch Feature Flag Evaluations. When using the Relay Proxy, make sure the URL you provide is publicly available. | `https://config.ff.harness.io/api/1.0` |

Reviewer: ```suggestion
| baseUrl | The URL used to fetch feature flag evaluations. When using the relay proxy, make sure the URL you provide is publicly available. | `https://config.ff.harness.io/api/1.0` |
```

> | eventUrl | The URL for posting metrics data to the Feature Flag service. When using the Relay Proxy, make sure the URL you provide is publicly available. | `https://events.ff.harness.io/api/1.0` |

Reviewer: ```suggestion
| eventUrl | The URL for posting metrics data to the feature flag service. When using the relay proxy, make sure the URL you provide is publicly available. | `https://events.ff.harness.io/api/1.0` |
```

| streamEnabled | Set to `true` to enable streaming mode. Set to `false` to disable streaming mode. | `true` |

```
options={{ // OPTIONAL: advanced options
        baseUrl: 'https://url-to-access-flags.com',
        eventUrl: 'https://url-for-events.com',
        streamEnabled: true,
        allAttributesPrivate: false,
        privateAttributeNames: ['customAttribute'],
        debug: true
      }}
```

> ## Evaluate a Flag

Reviewer: ```suggestion
## Evaluate a flag
```

> Evaluating a Flag is when the SDK processes all Flag rules and returns the correct Variation of that Flag for the Target you provide. 

Reviewer: ```suggestion
Evaluating a flag is when the SDK processes all flag rules and returns the correct variation of that flag for the target you provide. 
```

> If a matching Flag can’t be found, or the SDK can’t remotely fetch Flags, the default value is returned. The default value is undefined, however you can optionally set the default value to something else, which is particularly useful if you plan to use async mode on start up. For more information about async mode, go to [Use async mode](#use-async-mode).

Reviewer: ```suggestion
If a matching flag can’t be found, or the SDK can’t remotely fetch flags, the default value is returned. The default value is undefined, however you can optionally set the default value to something else, which is particularly useful if you plan to use async mode on startup. For more information about async mode, go to [Use async mode](#use-async-mode).
```

> With this SDK you can evaluate a single Flag or a multiple Flags. To do this, import and use one of the following hooks and pass in:

Reviewer: ```suggestion
With this SDK you can evaluate a single flag or a multiple flags. To do this, import and use one of the following hooks and pass in:
```

> * Identifier of the Flag you want to evaluate

Reviewer: ```suggestion
* Identifier of the flag you want to evaluate
```

> * (Optional) The default Variation 

Reviewer: ```suggestion
* (Optional) The default variation 
```

:::

### Display a loading component until Flags are returned in async mode

When using async mode and the `ifFeatureFlag HOC`, by default the component won’t display until the Flags are retrieved. This behavior can be overridden by passing a `loadingFallback` element in the options object. This will display the `loadingFallback` until the Flags are retrieved, then the component will either show or hide as normal. 

For example: 

```
import { ifFeatureFlag } from '@harnessio/ff-react-client-sdk'
 
// ...
 
function MyComponent() {
  return <p>This should render if the flag is on</p>
}
 
const MyConditionalComponent = ifFeatureFlag('flag1', {
  loadingFallback: <p>Loading...</p>
})(MyComponent)
```
#### Conditionally render a component

You can conditionally render a component by using `ifFeatureFlag`, which is a [higher-order component (HOC)](https://reactjs.org/docs/higher-order-components.html) that wraps your component and returns it only when the named Flag is enabled or matches a specific value. 

> ##### Render if the Flag is enabled 

Reviewer: ```suggestion
##### Render if the flag is enabled 
```

> By default `ifFeatureFlag` will render the component if the named Flag evaluates as truthy. In the below example, you can use `MyConditionalComponent` as a normal component and it will render the `MyComponent` component if the Flag is enabled.

Reviewer: truthy?

```
import { ifFeatureFlag } from '@harnessio/ff-react-client-sdk'
// ...
function MyComponent() {
  return <p>This renders if the flag is enabled</p>
}
const MyConditionalComponent = ifFeatureFlag('flag1')(MyComponent)
``` 
##### Render if the condition is matched

> By default the component will display if the Flag value is evaluated as truthy. This behavior can be overridden by passing the matchValue option in the options object. The Flag value will then be compared with `matchValue` to determine whether it should display. For example:

Reviewer: ```suggestion
By default the component will display if the flag value is evaluated as truthy. This behavior can be overridden by passing the matchValue option in the options object. The flag value will then be compared with `matchValue` to determine whether it should display. For example:
```

> By default the component will display if the Flag value is evaluated as truthy. This behavior can be overridden by passing the matchValue option in the options object. The Flag value will then be compared with `matchValue` to determine whether it should display. For example:

Reviewer: truthy?
